### PR TITLE
Add support for second-opinion

### PR DESCRIPTION
**Pull request by gene1wood**
This adds a new Nginx server which supports the
[second-opinion](https://github.com/gene1wood/second-opinion) model
which includes a new lua "layer" and a new options file.

### CODE REVIEW

File: webserver_configurations/OpenID_Connect/Nginx/conf.d/second_opinion/openidc_second_opinion_layer.lua [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+if f == nil then
+  ngx.log(ngx.ERR, "can't initialize loadfile: "..e)
+end
+ok, e = pcall(f)
+if not ok then
+  ngx.log(ngx.ERR, "can't load configuration: "..e)
+end
+
+-- Authenticate with the primary OP if necessary
+local res, err, url, session = oidc.authenticate(primary.opts)
```
Comment (gdestuynder) [inline webserver_configurations/OpenID_Connect/Nginx/conf.d/second_opinion/openidc_second_opinion_layer.lua]: session should directly be primary.session

File: webserver_configurations/OpenID_Connect/Nginx/conf.d/second_opinion/openidc_second_opinion_layer.lua [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+end
+
+ngx.log(ngx.DEBUG, "Primary authentication completed")
+
+-- Pass the user identity from the primary OP to the secondary OP
+secondary.opts.authorization_params = secondary.opts.authorization_params and secondary.opts.authorization_params or {}
+secondary.opts.authorization_params.login_hint = primary.session.data.id_token.email
+
+-- Authenticate with the secondary OP if necessary
+local res, err, url, session = oidc.authenticate(secondary.opts, nil, secondary.session_opts)
```
Comment (gdestuynder) [inline webserver_configurations/OpenID_Connect/Nginx/conf.d/second_opinion/openidc_second_opinion_layer.lua]: same as previous comment, 
```
local res, err, url
res, err, url, secondary.session = ...
```

File: webserver_configurations/OpenID_Connect/Nginx/conf.d/second_opinion/openidc_second_opinion_layer.lua [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+local function compare_sessions(s1, s2)
+  -- check that both OPs authenticated the same user
+  local res = s1.data.user.email == s2.data.user.email
+  if res == false then
+    ngx.log(ngx.ERR, "User authenticated as different identities in primary and secondary auth")
+  end
+  return res
+end
+
```
Comment (gdestuynder) [inline webserver_configurations/OpenID_Connect/Nginx/conf.d/second_opinion/openidc_second_opinion_layer.lua:80]: add comment to make it clear how this works, such as:
"we only want the RP to know about groups that we are sure are in both places"

File: webserver_configurations/OpenID_Connect/Nginx/conf.d/second_opinion/openidc_second_opinion_layer.lua [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  ngx.redirect(primary.opts.logout_path)
+end
+
+-- Replace the group list with the intersection of the group lists returned from each OP
+local groups = get_group_intersection(primary.session, secondary.session)
+primary.session.data.user.groups = groups
+primary.session.data.id_token.groups = groups
+
+-- Access control: only allow specific users in (this is optional, without it all authenticated users are allowed in)
+-- (TODO: add example)
```
Comment (gdestuynder) [inline webserver_configurations/OpenID_Connect/Nginx/conf.d/second_opinion/openidc_second_opinion_layer.lua]: Im guessing that in this very case we do actually want an example

**Review by gdestuynder — CHANGES_REQUESTED** [review]

File: webserver_configurations/OpenID_Connect/Nginx/conf.d/second_opinion/openidc_second_opinion_layer.lua (hunk 3 above)
Comment (gene1wood) [reply webserver_configurations/OpenID_Connect/Nginx/conf.d/second_opinion/openidc_second_opinion_layer.lua:80]: @gdestuynder This section checks to validate that the user which authenticated to each OP is the same user (comment is `check that both OPs authenticated the same user`)

Was this note just meant for the section of the code that finds the intersection of the `groups` returned by each OP?

**Comment by gene1wood** [comment]
@gdestuynder I've made the changes you suggest, let me know what you think.

File: webserver_configurations/OpenID_Connect/Nginx/conf.d/second_opinion/openidc_second_opinion_layer.lua (hunk 3 above)
Comment (gdestuynder) [reply webserver_configurations/OpenID_Connect/Nginx/conf.d/second_opinion/openidc_second_opinion_layer.lua:80]: whichever section at which you think its clearer

**Review by gdestuynder — APPROVED** [review]

**Pull request merged**